### PR TITLE
Add support for empty frames in k PRECEDING/FOLLOWING window frame bounds

### PR DESCRIPTION
**Pull request by pramodsatya**
Add support for empty frames in window frames with k PRECEDING and 
k FOLLOWING frame bounds in ROWS mode.
k can be an integer constant or a column containing integer values.

### CODE REVIEW

File: velox/core/PlanNode.h [hunk 1]
```diff
@@ -1487,9 +1487,11 @@ class WindowNode : public PlanNode {
   struct Frame {
     WindowType type;
     BoundType startType;
-    TypedExprPtr startValue;
+    std::optional<bool> isStartBoundConstant;
```
Comment (aditi-pandit) [inline velox/core/PlanNode.h:1490]: The older representation of TypedExprPtr is reasonable in PlanNode. The Window operator can distinguish between a constant and column expr (see similar code for constant function parameters).

File: velox/exec/Window.cpp [hunk 2]
```diff
@@ -41,16 +41,78 @@ void initKeyInfo(
...
+    vector_size_t currentRow,
+    std::optional<vector_size_t> offsetValue = 0) {
+  if (!offsetValue || type != core::WindowNode::WindowType::kRows) {
+    VELOX_FAIL("k preceding as frame end is allowed only in ROWS mode");
+  }
+  return std::min(
+      currentRow + vector_size_t(offsetValue.value()), partitionEndRow);
+}
+
+void setWindowFrameStart(
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:82]: This function can return a WindowFrameFunctionPtr instead of passing by reference.

File: velox/exec/Window.cpp [hunk 3]
```diff
@@ -41,16 +41,78 @@ void initKeyInfo(
...
-  VELOX_CHECK_EQ(windowFunction.frame.startValue, nullptr);
-  VELOX_CHECK_EQ(windowFunction.frame.endValue, nullptr);
+vector_size_t findFrameEndPointKPreceding(
+    core::WindowNode::WindowType type,
+    vector_size_t functionNumber,
+    vector_size_t partitionStartRow,
+    vector_size_t /*partitionEndRow*/,
+    vector_size_t currentRow,
+    std::optional<vector_size_t> offsetValue = 0) {
+  if (!offsetValue || type != core::WindowNode::WindowType::kRows) {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:51]: This shouldn't be a per-row check. Move it to initialization time.

File: velox/exec/Window.cpp [hunk 4]
```diff
@@ -313,6 +459,26 @@ void Window::callApplyForPartitionRows(
...
   }
 
+  // Compute bound values for frames with k preceding and k following bounds.
+  VectorPtr precedingVector = nullptr;
+  VectorPtr followingVector = nullptr;
+  FlatVector<int>* precedingFlatVector = nullptr;
+  FlatVector<int>* followingFlatVector = nullptr;
+  vector_size_t startOffset = 0;
+  vector_size_t endOffset = 0;
+  bool hasOffsetBasedBounds = false;
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:469]: Are you using hasOffsetBasedBounds variable anywhere ?

File: velox/exec/Window.cpp [hunk 5]
```diff
@@ -345,8 +511,54 @@ void Window::callApplyForPartitionRows(
     // as WindowFunction only sees one partition at a time.
     rawPeerStarts[j] = peerStartRow_ - firstPartitionRow;
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
+    if (precedingFlatVector) {
+      startOffset = precedingFlatVector->valueAt(j);
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:515]: Maybe you should initialize precedingFlatVector as a constant vector of value 0 and always get startOffset from that vector instead of doing a check for the pointer each time.

File: velox/core/PlanNode.h (hunk 1 above)
Comment (pramodsatya) [reply velox/core/PlanNode.h:1490]: Thanks for the suggestion, I was meaning to go through the constant parameter changes and modify accordingly, will update PR shortly.

File: velox/exec/Window.cpp (hunk 3 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp]: Thanks for the suggestion, removed these redundant checks.

File: velox/exec/Window.cpp (hunk 4 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp]: Removed it now, thanks!

File: velox/exec/Window.cpp (hunk 5 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp]: Thanks for the suggestion, will incorporate this in the PR for variable k values; removed precedingVector from this PR.

File: velox/exec/Window.cpp [hunk 6]
```diff
@@ -41,16 +41,94 @@ void initKeyInfo(
...
+  return partitionStartRow;
+}
+
+vector_size_t findFrameEndPointKPreceding(
+    core::WindowNode::WindowType type,
+    vector_size_t functionNumber,
+    vector_size_t partitionStartRow,
+    vector_size_t /*partitionEndRow*/,
+    vector_size_t currentRow,
+    std::optional<vector_size_t> offsetValue) {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:60]: Why is this optional ? I think you did this to make the signatures consistent across the functions. In that case please add a debug assertion that offset is always passed.

File: velox/exec/Window.cpp [hunk 7]
```diff
@@ -147,14 +225,54 @@ void Window::createWindowFunctions(
...
          windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+         std::nullopt,
+         std::nullopt});
+    if (hasKBasedStartBound) {
+      checkFrameType(
+          "k preceding/following as frame start is allowed only in ROWS mode");
+      auto frameStartValue = windowNodeFunction.frame.startValue;
+      if (auto constant = constantArg(frameStartValue)) {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:257]: Please can you check if k is always constant or can be a column reference also.

File: velox/exec/Window.cpp [hunk 8]
```diff
@@ -376,15 +548,38 @@ void Window::callApplyForPartitionRows(
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
 
     for (auto w = 0; w < numFuncs; w++) {
-      auto frameEndPoints = findFrameEndPoints(
-          w,
+      std::optional<vector_size_t> startOffset = std::nullopt;
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:551]: It might be better to determine the frameOffsets in the findFrameEndPointkPreceding(Following) functions, rather than introduce if statements here.

File: velox/exec/Window.cpp [hunk 9]
```diff
@@ -376,15 +548,38 @@ void Window::callApplyForPartitionRows(
...
+      rawFrameEndBuffers[w][j] =
+          windowFunctions_[w]->getFrameEndBoundFunction()(
+              windowFrames_[w].type,
+              w,
+              firstPartitionRow,
+              lastPartitionRow,
+              i,
+              endOffset) -
+          firstPartitionRow;
+      emptyFrameCheck(
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:578]: Why are you invoking emptyFrameCheck at this point ? We logically cannot have empty frame is many of the frame combinations. Since emptyFrameCheck is dependent of particular frame combinations maybe you can use the function pointer approach here also.

File: velox/exec/Window.cpp [hunk 10]
```diff
@@ -294,18 +417,52 @@ void Window::callResetPartition(vector_size_t partitionNumber) {
...
-  return std::make_pair(partitionStartRow, peerEndRow);
+void Window::computeVariableFrameBounds(
+    std::vector<vector_size_t>& constantPrecedingFrameOffset,
+    std::vector<vector_size_t>& constantFollowingFrameOffset) {
+  auto updateFrameVector =
+      [&](vector_size_t idx,
+          FrameArg frameArg,
+          std::vector<vector_size_t>& constantFrameOffset) -> void {
+    if (frameArg.constantValue) {
+      auto constantOffset =
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:428]: This value can be determined at initialization time itself and made a member variable to use here directly.

Comment (pramodsatya) [reply velox/exec/Window.cpp:428]: Thanks for the suggestion, added it as a parameter to windowFrame struct.

File: velox/exec/Window.cpp (hunk 8 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp:551]: Removed these checks, thanks!

File: velox/exec/Window.cpp (hunk 9 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp:536]: Thanks for the suggestion but I am not sure how the function pointer approach could be used here, could you please elaborate? I tried moving this to a class variable and removing this check but couldn't do so as the emptyFrameCheck is to be done after the rawFrameBuffers are computed for this index.

File: velox/exec/Window.cpp (hunk 6 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp:60]: Yes, that is why the optional param was added, included an assertion in this function too, but it is not required since a check is already there in createWindowFunctions.

File: velox/exec/Window.h [hunk 11]
```diff
@@ -22,6 +22,16 @@
 
 namespace facebook::velox::exec {
 
+// Structure for k preceding/following frame arguments, constantValue stores
+// the value of k for constant k and the column index when k is a column. For
+// columnar values of k, the columnarOffset stores the offsets for the rows
+// being parsed in the call to 'apply'.
+struct FrameArg {
+  bool isConstantValue;
```
Comment (aditi-pandit) [inline velox/exec/Window.h:30]: I feel its better to have 2 fields : (optional) constantValue and constantIndexValue (this can be kConstantChannel) instead of the bool. I didn't follow the explanation for columnOffset-- this is the vector of values from the frame channel ?

File: velox/exec/Window.cpp [hunk 12]
```diff
@@ -137,12 +137,63 @@ void Window::createWindowFunctions(
         operatorCtx_->pool(),
         &stringAllocator_));
 
-    windowFrames_.push_back(
+    auto updateKBasedFrameBounds = [&](const core::TypedExprPtr& frame,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:140]: It will be more readable if this function just returned the FrameArg and the caller assigned it correctly to the start or end frameArg

File: velox/exec/Window.h [hunk 13]
```diff
@@ -67,6 +77,8 @@ class Window : public Operator {
     const core::WindowNode::BoundType endType;
     const std::optional<column_index_t> startChannel;
```
Comment (aditi-pandit) [inline velox/exec/Window.h:78]: Are startChannel and endChannel needed anymore after your change for startFrameArg and endFrameArg ? If not, then changing their types from std::optional<column_index_t> to FrameArg should be sufficient.

File: velox/exec/Window.cpp [hunk 14]
```diff
@@ -379,25 +450,71 @@ void Window::callApplyForPartitionRows(
         }
         break;
       }
-      case core::WindowNode::BoundType::kPreceding:
-      case core::WindowNode::BoundType::kFollowing:
-        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kPreceding: {
+        kBoundFrameCheck(frameArg);
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:454]: This shouldn't be checked each time. I think we can assume that it was correctly set by the code.

File: velox/exec/Window.cpp [hunk 15]
```diff
@@ -350,10 +402,29 @@ void Window::callApplyForPartitionRows(
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
   }
 
+  auto kBoundFrameCheck = [&](std::optional<FrameArg> frameArg) -> void {
+    if (!frameArg) {
+      VELOX_FAIL("Frame argument is missing");
+    }
+  };
+
+  auto updateFrameVector = [&](FrameArg& frameArg) -> void {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:411]: Maybe call this readFrameValues ?

File: velox/exec/Window.cpp [hunk 16]
```diff
@@ -379,25 +450,71 @@ void Window::callApplyForPartitionRows(
...
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  auto emptyFrameCheck = [&](vector_size_t* frameStart,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:490]: Call this function fixEmptyFrames. Please also explain why this code is needed. Can we possibly do this check only for a subset of the frames ? If yes, then maybe we should do it only conditionally.

File: velox/exec/Window.h (hunk 13 above)
Comment (pramodsatya) [reply velox/exec/Window.h]: Apologies for missing this, I thought they were already removed.

File: velox/exec/Window.h (hunk 11 above)
Comment (pramodsatya) [reply velox/exec/Window.h]: Thanks for the suggestion, I will use these 2 fields and remove the bool. columnOffset is a FlatVector that stores the value in the column at the given index and is populated for numRows in the call to apply().

File: velox/exec/Window.cpp [hunk 17]
```diff
@@ -137,12 +137,61 @@ void Window::createWindowFunctions(
...
         {windowNodeFunction.frame.type,
          windowNodeFunction.frame.startType,
          windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+         std::nullopt,
+         std::nullopt});
+
+    if (hasKBasedStartBound) {
+      auto frameStartValue = windowNodeFunction.frame.startValue;
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:177]: It doesn't seem like you are using this variable anywhere. Same in the next if statement. Please remove them.

File: velox/exec/Window.cpp [hunk 18]
```diff
@@ -137,12 +137,61 @@ void Window::createWindowFunctions(
...
+        auto constantOffset =
+            constant->template as<ConstantVector<int64_t>>()->valueAt(0);
+        VELOX_USER_CHECK_GE(
+            constantOffset, 1, "k in frame bounds must be at least 1");
+        return FrameArg({constantOffset, kConstantChannel, nullptr});
+      } else {
+        FlatVectorPtr<int64_t> offsets =
+            std::dynamic_pointer_cast<FlatVector<int64_t>>(
+                BaseVector::create(BIGINT(), 0, operatorCtx_->pool()));
+        return FrameArg(
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:153]: You don't have to explicitly invoke the constructor. Just returning the tuple should suffice.

File: velox/exec/Window.h [hunk 19]
```diff
@@ -22,6 +22,16 @@
 
 namespace facebook::velox::exec {
 
+// Structure for k preceding/following frame arguments, constantValue stores
+// the value of k for constant k and constantIndexValue stores the index of the
+// column when k is a column. When k is a column, columnarOffset stores the
+// offsets for the rows being parsed in the call to 'apply'.
+struct FrameArg {
```
Comment (aditi-pandit) [inline velox/exec/Window.h:29]: This code can be simplified a bit. In this struct you can have column_index_t index and VectorPtr<int64_t> value. The VectorPtr can be a ConstantVector when k is a constant. index is kConstantChannel when k is a constant. This would simplify your code.

Comment (aditi-pandit) [inline velox/exec/Window.h:29]: Maybe call this struct FrameChannelArg ?

File: velox/exec/Window.cpp [hunk 20]
```diff
@@ -379,25 +442,73 @@ void Window::callApplyForPartitionRows(
...
     }
   };
 
+  auto fixOutOfBoundFrames = [&](vector_size_t* frameStart,
+                                 vector_size_t* frameEnd) -> void {
+    for (auto i = frameStart, j = frameEnd; i < frameStart + numRows;
+         i++, j++) {
+      if (*i <= *j) {
+        *i = std::max(*i, 0);
+        *j = std::min(*j, lastPartitionRow - firstPartitionRow);
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:486]: lastPartitionRow - firstPartitionRow can be computed once and passed to this function. It is the last row of the partition.

File: velox/exec/Window.cpp [hunk 21]
```diff
@@ -379,25 +442,73 @@ void Window::callApplyForPartitionRows(
...
+      } else {
+        *i = -1;
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:488]: Is this a convention to use -1 ? Is it similar in java code ?

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 22]
```diff
@@ -217,12 +256,14 @@ class AggregateWindowFunction : public exec::WindowFunction {
     // the previous and current row.
     for (int i = 0; i < numRows; i++) {
       auto currentFrameEnd = rawFrameEnds[i] - startFrame + 1;
-      if (currentFrameEnd > prevFrameEnd) {
-        computeAggregate(rows, prevFrameEnd, currentFrameEnd);
+      if (rawFrameStarts[i] == -1 && rawFrameEnds[i] == -1) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:259]: Based on the incremental logic we should never have empty frames that satisfy those conditions right ? So do we need this check ?

File: velox/exec/Window.cpp (hunk 18 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp:153]: Thanks for the suggestion, simplified accordingly!

File: velox/exec/Window.h [hunk 23]
```diff
@@ -22,6 +22,16 @@
 
 namespace facebook::velox::exec {
 
+// Structure for k preceding/following frame arguments, index stores the index
+// of the column when k is a column and kConstantChannel for constant k. When k
+// is a column, value points to a FlatVector containing the columnar offsets,
+// when k is a constant integer, value points to a ConstantVector of length 1
+// containing the constant offset.
+struct FrameChannelArg {
+  int64_t index;
```
Comment (aditi-pandit) [inline velox/exec/Window.h:31]: Please use column_index_t type for this.

File: velox/exec/Window.h [hunk 24]
```diff
@@ -65,8 +75,8 @@ class Window : public Operator {
     const core::WindowNode::WindowType type;
     const core::WindowNode::BoundType startType;
     const core::WindowNode::BoundType endType;
-    const std::optional<column_index_t> startChannel;
-    const std::optional<column_index_t> endChannel;
+    std::optional<FrameChannelArg> startFrameArg;
```
Comment (aditi-pandit) [inline velox/exec/Window.h:78]: Why did you remove the const qualifier here ?

File: velox/exec/Window.cpp [hunk 25]
```diff
@@ -350,10 +397,20 @@ void Window::callApplyForPartitionRows(
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
   }
 
-  auto updateFrameBounds = [&](vector_size_t* rawFrameBounds,
-                               core::WindowNode::BoundType boundType,
-                               core::WindowNode::WindowType type,
-                               bool isStartBound) -> void {
+  auto readFrameValues = [&](FrameChannelArg& frameArg) -> void {
+    if (frameArg.index != kConstantChannel) {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:401]: This check isn't needed.

File: velox/exec/Window.cpp [hunk 26]
```diff
@@ -350,10 +397,20 @@ void Window::callApplyForPartitionRows(
...
+      frameArg.value->resize(numRows);
+      windowPartition_->extractColumn(
+          frameArg.index, partitionOffset_, numRows, 0, frameArg.value);
+    }
+  };
+
+  auto updateFrameBounds =
+      [&](vector_size_t* rawFrameBounds,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:409]: The input parameters appear before output parameters in the argument list. A good order here would be windowType, boundType, isStartBound, frameArg, rawFrameBounds.

File: velox/exec/Window.cpp [hunk 27]
```diff
@@ -350,10 +397,20 @@ void Window::callApplyForPartitionRows(
...
+          core::WindowNode::BoundType boundType,
+          core::WindowNode::WindowType type,
+          bool isStartBound,
+          std::optional<FrameChannelArg> frameArg) -> void {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:413]: Should this be const & ?

File: velox/exec/Window.cpp [hunk 28]
```diff
@@ -379,25 +436,85 @@ void Window::callApplyForPartitionRows(
...
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  auto fixOutOfBoundFrames = [&](vector_size_t* frameStart,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:482]: Please add a comment about what this code does.

File: velox/exec/Window.cpp [hunk 29]
```diff
@@ -137,12 +137,58 @@ void Window::createWindowFunctions(
...
+         windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kFollowing);
+
+    WindowFrame windowFrame(
         {windowNodeFunction.frame.type,
          windowNodeFunction.frame.startType,
          windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+         std::nullopt,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:172]: The frameArgs can be constructed in place 

hasKBasedStartBound ? updateKBasedFrameBounds(....) : std::nullopt

File: velox/exec/Window.cpp [hunk 30]
```diff
@@ -379,25 +436,85 @@ void Window::callApplyForPartitionRows(
...
+        true,
+        windowFrames_[i].startFrameArg);
     updateFrameBounds(
         rawFrameEndBuffers[i],
         windowFrames_[i].endType,
         windowFrames_[i].type,
-        false);
+        false,
+        windowFrames_[i].endFrameArg);
+  }
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:510]: It might be simpler to add the fixOutOfBoundFrames in this loop itself. Seems like that check is applicable only if either bound type has kPreceding or kFollowing. So if either had a FrameChannelArg then the out of bound check is applicable.

The hasOutOfBoundFrames_ boolean is not needed.

File: velox/core/PlanNode.cpp [hunk 31]
```diff
@@ -504,6 +504,18 @@ void addWindowFunction(
   if (frame.endType == WindowNode::BoundType::kUnboundedPreceding) {
     VELOX_USER_FAIL("Window frame end cannot be UNBOUNDED PRECEDING");
   }
+  if (frame.type == WindowNode::WindowType::kRange) {
+    if (frame.startType == WindowNode::BoundType::kPreceding ||
+        frame.startType == WindowNode::BoundType::kFollowing) {
+      VELOX_USER_FAIL(
```
Comment (aditi-pandit) [inline velox/core/PlanNode.cpp:510]: Should this be VELOX_NYI and not VELOX_USER_FAIL. We should implement kPreceding/kFollowing for RANGE mode in Velox.

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 32]
```diff
@@ -155,21 +164,45 @@ class AggregateWindowFunction : public exec::WindowFunction {
     // computation for each row needs the values from its frameStart to frameEnd
     // indices. So for each apply() we read arguments from the least
     // frameStart to the highest frameEnd value in the row block.
-    vector_size_t firstRow = rawFrameStarts[0];
-    vector_size_t lastRow = rawFrameEnds[0];
+    vector_size_t i = 0;
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:167]: https://github.com/facebookincubator/velox/pull/3200 is submitted now. Please can you rebase.

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 33]
```diff
@@ -216,13 +250,17 @@ class AggregateWindowFunction : public exec::WindowFunction {
     // incrementally aggregate over the new rows seen in the frame between
     // the previous and current row.
     for (int i = 0; i < numRows; i++) {
-      auto currentFrameEnd = rawFrameEnds[i] - startFrame + 1;
-      if (currentFrameEnd > prevFrameEnd) {
-        computeAggregate(rows, prevFrameEnd, currentFrameEnd);
+      if (isEmptyFrame(rawFrameStarts[i], rawFrameEnds[i])) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:253]: Shouldn't empty frame never occur in the incremental Aggregation case ?

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 34]
```diff
@@ -275,6 +317,15 @@ class AggregateWindowFunction : public exec::WindowFunction {
...
   vector_size_t singleGroupRowSize_;
 
+  // This offset tracks how far along the partition rows have been output.
+  // This can be used to optimize reading offset column values corresponding
+  // to the present row set in getOutput.
+  vector_size_t partitionOffset_;
+  // Tracks if a frame bound of 'between unbounded preceding and k preceding' is
+  // seen in the aggregate function call. This is used to adjust the number of
+  // rows read to compute the aggregate.
+  bool hasEmptyPrecedingFrame_;
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:327]: Why is this a member variable ?

File: velox/exec/Window.cpp (hunk 27 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp]: Thanks for the suggestion!

File: velox/exec/Window.cpp (hunk 26 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp]: Thanks for the suggestion but since the frameArg is optional, it is retained as the last parameter. The rawFrameBounds which is the output is moved to the penultimate position.

File: velox/exec/Window.cpp (hunk 30 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp]: Thanks for the suggestion, it was in a separate loop to avoid checking in the loop for other frames. Changed accordingly!

File: velox/exec/Window.h (hunk 23 above)
Comment (pramodsatya) [reply velox/exec/Window.h]: Apologies for missing this. Changed to column_index_t.

File: velox/exec/Window.h (hunk 24 above)
Comment (pramodsatya) [reply velox/exec/Window.h]: Thanks for pointing it out, it was removed because an empty FrameArg object was being set from updateKBasedFrameBounds before, reintroduced const.

File: velox/functions/prestosql/window/AggregateWindow.cpp (hunk 33 above)
Comment (pramodsatya) [reply velox/functions/prestosql/window/AggregateWindow.cpp:296]: Thanks for pointing it out, this is needed for cases like "rows between unbounded preceding and k preceding" where empty frame can occur.

File: velox/functions/prestosql/window/AggregateWindow.cpp (hunk 34 above)
Comment (pramodsatya) [reply velox/functions/prestosql/window/AggregateWindow.cpp:369]: When an empty preceding frame is seen in the first call to apply, the startRow and numFrameRows should be changed accordingly in all subsequent calls for correctness of the aggregate calculation.

File: velox/core/PlanNode.cpp [hunk 35]
```diff
@@ -504,6 +504,18 @@ void addWindowFunction(
   if (frame.endType == WindowNode::BoundType::kUnboundedPreceding) {
     VELOX_USER_FAIL("Window frame end cannot be UNBOUNDED PRECEDING");
   }
+  if (frame.type == WindowNode::WindowType::kRange) {
+    if (frame.startType == WindowNode::BoundType::kPreceding ||
+        frame.startType == WindowNode::BoundType::kFollowing) {
+      VELOX_NYI(
+          "k preceding/following as frame start is allowed only in ROWS mode");
```
Comment (aditi-pandit) [inline velox/core/PlanNode.cpp:511]: Its sufficient to say that "k preceding/following is only supported in ROWS mode" and put all the cases in a single condition.

File: velox/exec/Window.cpp [hunk 36]
```diff
@@ -137,12 +137,49 @@ void Window::createWindowFunctions(
         operatorCtx_->pool(),
         &stringAllocator_));
 
-    windowFrames_.push_back(
-        {windowNodeFunction.frame.type,
-         windowNodeFunction.frame.startType,
-         windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+    auto updateKBasedFrameBounds = [&](const core::TypedExprPtr& frame,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:140]: Maybe call this createFrameChannelArg since there isn't any update per se.

File: velox/exec/Window.cpp [hunk 37]
```diff
@@ -137,12 +137,49 @@ void Window::createWindowFunctions(
...
+                                       WindowFrame& windowFrame,
+                                       bool isStartFrame) -> FrameChannelArg {
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:142]: Doesn't seem like isStartFrame is used in the function. Please remove it.

File: velox/exec/Window.cpp [hunk 38]
```diff
@@ -137,12 +137,49 @@ void Window::createWindowFunctions(
...
         &stringAllocator_));
 
-    windowFrames_.push_back(
-        {windowNodeFunction.frame.type,
-         windowNodeFunction.frame.startType,
-         windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+    auto updateKBasedFrameBounds = [&](const core::TypedExprPtr& frame,
+                                       WindowFrame& windowFrame,
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:141]: const &

File: velox/exec/Window.cpp [hunk 39]
```diff
@@ -137,12 +137,49 @@ void Window::createWindowFunctions(
...
+                                       bool isStartFrame) -> FrameChannelArg {
+      if (auto constant = constantArg(frame)) {
+        auto constantOffset =
+            constant->template as<ConstantVector<int64_t>>()->valueAt(0);
+        VELOX_USER_CHECK_GE(
+            constantOffset, 1, "k in frame bounds must be at least 1");
+        return {kConstantChannel, constant};
+      } else {
+        FlatVectorPtr<int64_t> offsets =
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:150]: The FlatVectorPtr casting is not needed. VectorPtr offsets can simply be BaseVector::create(...).

File: velox/exec/Window.cpp [hunk 40]
```diff
@@ -379,25 +425,88 @@ void Window::callApplyForPartitionRows(
...
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  // Updates the start and end frame bounds to a lower bound of 0 and upper
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:471]: Nit : Rephrase
"k preceding and k following bounds in ROWS mode can go over the partition limits. So they are bound to the first and last partition rows if required. If the frame is empty incase the frameEnd > frameStart values, then the bounds are set to -1. "

File: velox/core/PlanNode.cpp (hunk 35 above)
Comment (aditi-pandit) [reply velox/core/PlanNode.cpp:511]: On second thoughts, do you want to put this error here. It might be better to put it only in the Window operator since the PlanBuilder tests wouldn't need to be modified,

File: velox/exec/Window.h [hunk 41]
```diff
@@ -22,6 +22,16 @@
 
 namespace facebook::velox::exec {
 
+// Structure for k preceding/following frame arguments, index stores the index
```
Comment (aditi-pandit) [inline velox/exec/Window.h:25]: Nit -- Rephrase 
"Used for k preceding/following frames. Index is the column index if k is a column. value is used to read the column values from the input, or contains the constant value if k is a constant."

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 42]
```diff
@@ -175,22 +179,44 @@ class AggregateWindowFunction : public exec::WindowFunction {
     bool usePreviousAggregate;
   };
 
+  bool isEmptyFrame(vector_size_t frameStart, vector_size_t frameEnd) {
+    return (frameStart == frameEnd) && (frameEnd == -1);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:183]: Its easier to read this as frameStart == -1 && frameEnd == -1

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 43]
```diff
@@ -175,22 +179,44 @@ class AggregateWindowFunction : public exec::WindowFunction {
     bool usePreviousAggregate;
   };
 
+  bool isEmptyFrame(vector_size_t frameStart, vector_size_t frameEnd) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:182]: This can be simply called emptyFrame

File: velox/functions/prestosql/window/NthValue.cpp [hunk 44]
```diff
@@ -131,7 +131,11 @@ class NthValueFunction : public exec::WindowFunction {
     auto frameStart = frameStarts[i];
     auto frameEnd = frameEnds[i];
     auto rowNumber = frameStart + offset - 1;
-    rowNumbers_[i] = rowNumber <= frameEnd ? rowNumber : -1;
+    if (frameStart == frameEnd && frameStart == -1) {
+      rowNumbers_[i] = -1;
+    } else {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/NthValue.cpp:136]: rowNumber can be initialized in the else part of the code.

File: velox/functions/prestosql/window/NthValue.cpp [hunk 45]
```diff
@@ -131,7 +131,11 @@ class NthValueFunction : public exec::WindowFunction {
     auto frameStart = frameStarts[i];
     auto frameEnd = frameEnds[i];
     auto rowNumber = frameStart + offset - 1;
-    rowNumbers_[i] = rowNumber <= frameEnd ? rowNumber : -1;
+    if (frameStart == frameEnd && frameStart == -1) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/NthValue.cpp:134]: This condition too is easier to read as frameStart == -1 && frameEnd == -1

File: velox/functions/prestosql/window/AggregateWindow.cpp [hunk 46]
```diff
@@ -175,22 +179,44 @@ class AggregateWindowFunction : public exec::WindowFunction {
...
       vector_size_t numRows) {
-    vector_size_t firstRow = rawFrameStarts[0];
+    vector_size_t i = 0;
+    while (i < numRows && rawFrameStarts[i] < 0 && rawFrameEnds[i] < 0) {
+      i++;
+    }
+    if (i) {
+      hasEmptyPrecedingFrame_ = true;
+    }
+    if (i == numRows) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/AggregateWindow.cpp:197]: In this case it seems like the result will be null for all rows, then why not optimize that path ?

File: velox/functions/prestosql/window/AggregateWindow.cpp (hunk 33 above)
Comment (aditi-pandit) [reply velox/functions/prestosql/window/AggregateWindow.cpp:296]: I don't entirely follow the optimizations for empty frames. Lets discuss in person.
